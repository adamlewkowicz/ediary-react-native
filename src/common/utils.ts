import dayjs from 'dayjs';
import { DateDay, UnitType } from '../types';
import { UNIT_TYPES } from './consts';

let timeout: NodeJS.Timeout;

export const debounce_ = () => {
  let timeout: NodeJS.Timeout;
  return (fn: () => void, delay = 250) => {
    clearTimeout(timeout);
    timeout = setTimeout(fn, delay);
  }
}

export const debounce = (fn: () => void, delay = 150) => {
  clearTimeout(timeout);
  timeout = setTimeout(fn, delay);
}

export const round = (value: number, scale = 10) => Math.round(value * scale) / scale;

export const toLocaleString = (value: number) => new Intl.NumberFormat('pl-PL').format(value);

export const mapAsyncSequence = async <T, R>(
  items: T[],
  callback: (item: T) => Promise<R>
) => {
  const result: R[] = [];

  for (const item of items) {
    const resultItem = await callback(item);
    result.push(resultItem);
  }

  return result;
}

export function getDayFromDate(
  date: dayjs.ConfigType
): DateDay {
  const dateDay: DateDay = dayjs(date).format('YYYY-MM-DD') as any;
  return dateDay;
}


export function getValAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const unit = UNIT_TYPES.find(unit => value.includes(unit));
  const parsedValue = parseFloat(
    value.replace(/,/, '.')
  );
  const result = Number.isNaN(parsedValue) ? null : parsedValue; 

  return {
    unit: unit || null,
    value: result || null
  }
}

export function sortByMostAccurateName(
  phrase: string
) {
  const phraseLowered = phrase.toLowerCase();

  return function<T extends SortByName>(
    a: T,
    b: T
  ): -1 | 0 | 1 {
    const aLowered = a.name.toLowerCase();
    const bLowered = b.name.toLowerCase();
  
    const aIndex = aLowered.indexOf(phraseLowered);
    const bIndex = bLowered.indexOf(phraseLowered);
  
    const notFoundIndex = -1;
    const orderByShorter = aLowered.length > bLowered.length ? 1 : -1;
  
    if (aIndex === notFoundIndex && bIndex === notFoundIndex) {
      return -1;
    }
    if (aIndex === notFoundIndex) {
      return 1;
    }
    if (bIndex === notFoundIndex) {
      return -1;
    }
  
    if (aLowered.length === bLowered.length) {
      return 0;
    }
  
    return orderByShorter;
  }
}

interface SortByName {
  name: string
}

export function getNumAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const parseToNumber = (val: string): number | null => {
    const parsed = val.trim().replace(/,/, '.')
    const numbers = parsed.match(/[+-]?\d+(\.\d+)?/g);

    if (!numbers) {
      return null;
    }
    const result = parseFloat(numbers[numbers.length - 1]);

    if (Number.isNaN(result)) {
      return null;
    }
    return result;
  }

  const unit = UNIT_TYPES.find(unit => value.includes(` ${unit}`)) || null;
  const result = parseToNumber(value);

  return {
    value: result,
    unit
  }
}

export function filterByUniqueId<T extends { id: number | string }>(
  item: T,
  index: number,
  self: T[]
): boolean {
  const foundIndex = self.findIndex(anyItem => anyItem.id === item.id);
  return foundIndex === index;
}

export function parseNumber(
  value: string,
  maxNumber?: number,
  maxLength?: number,
): string {
  const trimmed = maxLength ? value.substring(0, maxLength) : value;
  const cleaned = trimmed.replace(/[^\d.-]/g, '');
  const parsed = Number(cleaned);

  if (maxNumber && parsed >= maxNumber) {
    return cleaned.slice(0, -1);
  }

  return cleaned;
}

export function findOrFail<T>(
  array: T[],
  matcher: (item: T, index: number, self: T[]) => boolean
) {
  const foundItem = array.find(matcher);
  if (!foundItem) {
    throw new Error(
      '[findOrFail] - Item could not be found for specified criteria'
    );
  }
  return foundItem;
}