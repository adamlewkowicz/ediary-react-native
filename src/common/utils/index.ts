/// <reference lib="dom" />
import dayjs from 'dayjs';
import { DateDay, UnitType, DateTime, MacroElements, BaseMacroElements, ObjectEntries } from '../../types';
import { UNIT_TYPES, DATE_TIME, DATE_DAY, MACRO_ELEMENTS, KCAL_IN_ONE_MACRO_GRAM } from '../consts';
import { LayoutAnimation } from 'react-native';

export const debounce = () => {
  let timeout: NodeJS.Timeout;
  return (callback: () => void, delay = 250) => {
    clearTimeout(timeout);
    timeout = setTimeout(callback, delay);
  }
}

export const round = (value: number, scale = 10) => Math.round(value * scale) / scale;

export const toLocaleString = (value: number) => new Intl.NumberFormat('pl-PL').format(value);

/**
 * Maps over an array and runs promise returned from a callback in a sequence.
 * @param items array of elements to iterate on
 * @param callback callback that is executed on each iteration
 */
export const mapAsyncSequence = async <T, R>(
  items: T[],
  callback: (item: T) => Promise<R>
) => {
  const result: R[] = [];

  for (const item of items) {
    const resultItem = await callback(item);
    result.push(resultItem);
  }

  return result;
}

export function getDayFromDate(
  date: dayjs.ConfigType
): DateDay {
  const dateDay: DateDay = dayjs(date).format(DATE_DAY) as any;
  return dateDay;
}

export function getTimeFromDate(
  date: dayjs.ConfigType
) {
  const dateTime: DateTime = dayjs(date).format(DATE_TIME) as any;
  return dateTime;
}

export function getValAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const unit = UNIT_TYPES.find(unit => value.includes(unit));
  const parsedValue = parseFloat(
    value.replace(/,/, '.')
  );
  const result = Number.isNaN(parsedValue) ? null : parsedValue; 

  return {
    unit: unit || null,
    value: result || null
  }
}

export function sortByMostAccurateName(
  phrase: string
) {
  const phraseLowered = phrase.toLowerCase();

  return function<T extends { name: string }>(
    a: T,
    b: T
  ): -1 | 0 | 1 {
    const aNameLowered = a.name.toLowerCase();
    const bNameLowered = b.name.toLowerCase();
  
    const aIndex = aNameLowered.indexOf(phraseLowered);
    const bIndex = bNameLowered.indexOf(phraseLowered);
  
    const notFoundIndex = -1;
    const orderByShorter = aNameLowered.length > bNameLowered.length ? 1 : -1;

    const aNameHasNotBeenFound = aIndex === notFoundIndex;
    const bNameHasNotBeenFound = bIndex === notFoundIndex;
  
    if (aNameHasNotBeenFound && bNameHasNotBeenFound) {
      return -1;
    }
    if (aNameHasNotBeenFound) {
      return 1;
    }
    if (bNameHasNotBeenFound) {
      return -1;
    }
  
    if (aNameLowered.length === bNameLowered.length) {
      return 0;
    }
  
    return orderByShorter;
  }
}

export function getNumAndUnitFromString(value: string): {
  value: number | null
  unit: UnitType | null
} {
  const parseToNumber = (val: string): number | null => {
    const parsed = val.trim().replace(/,/, '.')
    const numbers = parsed.match(/[+-]?\d+(\.\d+)?/g);

    if (!numbers) {
      return null;
    }
    const result = parseFloat(numbers[numbers.length - 1]);

    if (Number.isNaN(result)) {
      return null;
    }
    return result;
  }

  const unit = UNIT_TYPES.find(unit => value.includes(` ${unit}`)) || null;
  const result = parseToNumber(value);

  return {
    value: result,
    unit
  }
}

export const filterByUniqueProperty = <
  E extends object,
  A extends E[]
>(property: keyof E) => (element: E, index: number, self: A): boolean => {
  const foundIndex = self.findIndex(anyElement =>
    anyElement[property] === element[property]
  );
  return foundIndex === index;
}

export const filterByUniqueId = filterByUniqueProperty('id');

export function parseNumber(
  value: string,
  maxNumber?: number,
  maxLength?: number,
): string {
  const trimmed = maxLength ? value.substring(0, maxLength) : value;
  const cleaned = trimmed.replace(/[^\d.-]/g, '');
  const parsed = Number(cleaned);

  if (maxNumber && parsed >= maxNumber) {
    return cleaned.slice(0, -1);
  }

  return cleaned;
}

export function findOrFail<T>(
  array: T[],
  matcher: (item: T, index: number, self: T[]) => boolean
): T {
  const foundItem = array.find(matcher);
  if (!foundItem) {
    throw new Error(
      '[findOrFail] - No item could be found for specified criteria'
    );
  }
  return foundItem;
}

const createArrayOfLength = <T>(
  length: number,
  callback: (index: number) => T
): T[] => Array.from({ length }, (_, index) => callback(index));

export const fillArrayWithinRange = (
  { from, to }: { from: number; to: number }
): number[] => {
  const zeroBasedCountFill = 1;
  const length = to - from + zeroBasedCountFill;
  return createArrayOfLength(length, index => index + from);
}

export const fetchify = async <T>(
  input: RequestInfo,
  init: RequestInit = {},
  controller = new AbortController()
): Promise<T> => {
  const { signal } = controller;

  const response = await fetch(input, { signal, ...init });
  const json: T = await response.json();

  return json;
}

export const objectMap = <
  T extends object,
  Property extends keyof T = keyof T,
  Value = T[Property]
>(
  obj: T,
  callback: (property: Property, value: Value) => Value
): T => {
  return Object.fromEntries(
    Object
      .entries(obj)
      .map(([property, value]) => [property, callback(property as Property, value)])
  ) as T;
}

const objectEntries: ObjectEntries = Object.entries;

export const calculateCaloriesByMacro = (
  macro: BaseMacroElements
): number => {
  return objectEntries(macro)
    .reduce(
      (kcal, [macroName, macroQuantity]) => kcal += macroQuantity * KCAL_IN_ONE_MACRO_GRAM[macroName],
      0
    );
}

export const layoutAnimateEase = (onAnimationDidEnd?: () => void) => {
  LayoutAnimation.configureNext(
    LayoutAnimation.Presets.easeInEaseOut,
    onAnimationDidEnd
  );
}

export const calculatePercentage = (portion: number, total: number): number => {
  if (total === 0) return 0;
  return Math.floor(portion / total * 100);
}

export const calculateMacroPerQuantity = <T extends MacroElements>(
  macroValuesPerHundredQuantity: T,
  quantity: number
): T => objectMap(
  macroValuesPerHundredQuantity,
  (_, macroValue: number) => round(macroValue * quantity / 100, 100)
);

export const calculateMacroPercentages = <T extends BaseMacroElements>(
  macroValues: T
): BaseMacroElements => {
  const macroSum = macroValues.carbs + macroValues.prots + macroValues.fats;

  const carbs = calculatePercentage(macroValues.carbs, macroSum);
  const prots = calculatePercentage(macroValues.prots, macroSum);
  const fats = calculatePercentage(macroValues.fats, macroSum);

  return { carbs, prots, fats };
}

const calculateMacroValueNeed = (
  eaten: number,
  needed: number
): MacroNeed => ({
  eaten,
  needed,
  left: needed - eaten,
  percentage: calculatePercentage(eaten, needed),
});

export const calculateMacroNeeds = <T extends object>(
  macroValuesEaten: T,
  macroValuesNeeded: T
): { [key: string]: MacroNeed } => objectMap(
  macroValuesEaten,
  (macroName, macroValueEaten) => {
    // TODO: refactor types
    const macroValueNeeded = macroValuesNeeded[macroName] as any as number;

    const macroValueNeed = calculateMacroValueNeed(
      macroValueEaten,
      macroValueNeeded
    );

    return macroValueNeed;
  }
);

type MacroNeed = {
  eaten: number
  needed: number
  left: number,
  percentage: number
}

const calculateObjectsValueSum = <T extends { [key: string]: number }>(
  objects: T[],
  initialObj: T = objects[0],
): T => {
  return objects.reduce<T>((acc, current) => {
    const next = { ...acc };

    for (const key in next) {
      // @ts-ignore
      next[key] += current[key];
    }

    return next;
  }, { ...initialObj });
}