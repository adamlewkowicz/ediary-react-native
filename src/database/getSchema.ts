import { getConnection, Connection } from 'typeorm';

export async function getDatabaseSchema(
  connection: Connection = getConnection()
): Promise<string[]> {
  const tableSchema: TableSchema[] = await connection.query(
    `SELECT * FROM sqlite_master WHERE type = "table";`
  );
  return tableSchema.map(schema => schema.sql);
}

interface TableSchema {
  name: string
  rootpage: number
  sql: string
  tbl_name: string
  type: 'table'
}