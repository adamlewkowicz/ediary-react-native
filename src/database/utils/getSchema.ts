import { getConnection, Connection } from 'typeorm';
import sqlFormatter from 'sql-formatter';

export async function getDatabaseSchema(
  connection: Connection = getConnection(),
  metaTableNames: string[] = ['sqlite_sequence', 'migrations']
): Promise<ExtendedTableSchema[]> {
  const tableSchema: TableSchema[] = await connection.query(
    `SELECT * FROM sqlite_master WHERE type = "table";`
  );
  const normalizedSchema: ExtendedTableSchema[] = tableSchema
    .filter(data => !metaTableNames.includes(data.tbl_name))
    .map(data => ({
      ...data,
      prettySql: sqlFormatter.format(data.sql)
    }))
    .sort((a, b) => a.tbl_name > b.tbl_name ? 1 : -1);
    
  return normalizedSchema;
}

interface ExtendedTableSchema extends TableSchema {
  prettySql: string
}
interface TableSchema {
  name: string
  rootpage: number
  sql: string
  tbl_name: string
  type: 'table'
}