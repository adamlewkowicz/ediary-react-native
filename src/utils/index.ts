export * from './fetchify';
export * from './calculate';
export * from './common';
export * from './generic';
export * from './parse';
export * from './sort';
export * from './date';
export * from './a11y';