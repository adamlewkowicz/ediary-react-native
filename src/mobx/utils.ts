import { Training, ExerciseSet, Exercise } from '../database/entities';
import { ExerciseSetState, ExerciseState, TrainingState } from './GymTrainingStore';

export const normalizeExerciseSet = (exerciseSet: ExerciseSet): ExerciseSetState => ({
  ...exerciseSet,
  duration: 0,
  isRest: false,
  restTime: 20,
  restDuration: 0,
  state: 'unfinished',
});

export const normalizeExercise = (exercise: Exercise) => {
  const { sets = [], ..._exercise } = exercise;

  const normalizedExerciseSets = sets.map(normalizeExerciseSet);

  const normalizedExercise: ExerciseState = {
    ..._exercise,
    duration: 0,
    setIds: normalizedExerciseSets.map(set => set.id),
  }

  return {
    exercise: normalizedExercise,
    exerciseSets: normalizedExerciseSets,
  }
}

const normalizeExercises = (exercises: Exercise[]) => {
  return exercises.reduce<NormalizedExercises>((normalized, _exercise) => {
    const { exercise, exerciseSets } = normalizeExercise(_exercise);

    return {
      exercises: [...normalized.exercises, exercise],
      exerciseSets: [...normalized.exerciseSets, ...exerciseSets]
    };
  }, { exercises: [], exerciseSets: [] });
}

export function normalizeTrainings(payload: Training[]) {
  return payload.reduce<NormalizedTrainings>((normalized, training) => {
    const { exercises = [], ...restTraining } = training;

    const result = normalizeExercises(exercises);

    const normalizedTraining: TrainingState = {
      ...restTraining,
      exerciseIds: exercises.map(exercise => exercise.id),
      isPaused: false,
      isActive: false,
    }

    return {
      trainings: [...normalized.trainings, normalizedTraining],
      exercises: [...normalized.exercises, ...result.exercises],
      exerciseSets: [...normalized.exerciseSets, ...result.exerciseSets],
    };
  }, { trainings: [], exercises: [], exerciseSets: [] });
}

interface NormalizedExercises {
  exercises: ExerciseState[]
  exerciseSets: ExerciseSetState[]
}

interface NormalizedTrainings extends NormalizedExercises  {
  trainings: TrainingState[]
}