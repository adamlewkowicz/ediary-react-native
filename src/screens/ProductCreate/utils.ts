import { Product } from '../../database/entities';
import * as Utils from '../../utils';

export const serializeProduct = (product?: Product): FormData => {
  const defaultValues = {
    name: '',
    brand: '',
    producer: '',
    portionQuantity: '',
    carbs: '',
    sugars: '',
    prots: '',
    fats: '',
    fattyAcids: '',
    kcal: '',
    barcode: '',
  };

  if (product) {
    const { macro, barcode, portion, name } = product;

    const parsedMacro = Object.fromEntries(
      Object.entries(macro).map(
        ([prop, value]) => [prop, String(value)]
      )
    );

    const parsedBarcode = Utils.isNil(barcode) ? '' : String(barcode);

    const formData: FormData = {
      ...defaultValues,
      ...parsedMacro,
      name,
      barcode: parsedBarcode,
      portionQuantity: String(portion),
    };

    return formData;
  }

  return defaultValues;
}

export const deserializeProduct = (productData: FormData) => {
  const { fattyAcids, sugars, ...restData } = productData;

  const macro = {
    carbs: Number(productData.carbs),
    prots: Number(productData.prots),
    fats: Number(productData.fats),
    kcal: Number(productData.kcal)
  }

  const barcode = productData.barcode.length ? productData.barcode : null;
  const portionQuantity = Number(productData.portionQuantity);

  const product = {
    ...restData,
    macro,
    barcode,
    portionQuantity,
  }

  return product;
}

export type FormData = {
  name: string
  brand: string
  producer: string
  portionQuantity: string
  carbs: string
  sugars: string
  prots: string
  fats: string
  fattyAcids: string
  kcal: string
  barcode: string
}