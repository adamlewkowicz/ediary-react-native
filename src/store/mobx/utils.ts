import { NormalizedTrainings, NormalizedExercises } from './types';
import { Training } from '../../database/entities';
import { TrainingState, ExerciseState, ExerciseSetState } from './training';
import { ExerciseStore } from './ExerciseStore';

export function normalizeTrainings(
  payload: Training[],
  exerciseSetRestTime = 5
): NormalizedTrainings {
  return payload.reduce((
    normalized: NormalizedTrainings,
    trainingEntity
  ): NormalizedTrainings => {
    const { exercises = [], ...training } = trainingEntity;
    
    const normalizedTraining: TrainingState = {
      ...training,
      isActive: false,
      isPaused: false,
      exerciseIds: exercises.map(exercise => exercise.id)
    }

    const {
      exercises: normalizedExercises,
      exerciseSets: normalizedExerciseSets
    } = exercises.reduce((
      normalized: NormalizedExercises,
      exerciseEntity,
    ): NormalizedExercises => {
      const { sets = [], ...exercise } = exerciseEntity;

      const normalizedExercise = {
        ...exercise,
        setIds: sets.map(set => set.id),
        duration: 0,
        entity: exerciseEntity,
      }
      const exerciseStore = new ExerciseStore(exerciseEntity);

      const normalizedSets: ExerciseSetState[] = sets.map(set => ({
        ...set,
        duration: 0,
        restTime: exerciseSetRestTime,
        restDuration: 0,
        isRest: false,
        state: 'unfinished'
      }));

      return {
        exercises: [...normalized.exercises, exerciseStore],
        exerciseSets: [...normalized.exerciseSets, ...normalizedSets]
      }
    }, { exercises: [], exerciseSets: [] });

    return {
      trainings: [...normalized.trainings, normalizedTraining],
      exercises: [...normalized.exercises, ...normalizedExercises],
      exerciseSets: [...normalized.exerciseSets, ...normalizedExerciseSets]
    }
  }, { trainings: [], exercises: [], exerciseSets: [] });
}