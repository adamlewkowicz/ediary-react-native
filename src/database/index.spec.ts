import { getDatabaseSchema } from './utils/getSchema';
import { getConnection } from 'typeorm';

describe('Database 🗄️', () => {

  describe('migrations', () => {

    it('should migrate without issues', async () => {
      const connection = getConnection();
      await connection.dropDatabase();
      await connection.runMigrations();
    });

    it('should revert migrations without issues', async () => {
      const connection = getConnection();
      await connection.dropDatabase();
      await connection.runMigrations();
    
      for (const _migration of connection.migrations) {
        await connection.undoLastMigration();
      }
    
      const tables = await getDatabaseSchema();
      expect(tables.length).toBe(0);
    });

  });

  describe('migrated schema', () => {

    it('should match snapshot', async () => {
      const connection = getConnection();
      await connection.dropDatabase();
      await connection.runMigrations();
  
      const schema = await getDatabaseSchema();
  
      for (const { prettySql, tbl_name } of schema) {
        expect(prettySql).toMatchSnapshot(tbl_name);
      }
    });

  });

  describe('automatically generated schema', () => {

    it('should match snapshot', async () => {
      const connection = getConnection();
      await connection.dropDatabase();
      await connection.synchronize();
      const schema = await getDatabaseSchema();
    
      for (const { prettySql, tbl_name } of schema) {
        expect(prettySql).toMatchSnapshot(tbl_name);
      }
    });

  });

});