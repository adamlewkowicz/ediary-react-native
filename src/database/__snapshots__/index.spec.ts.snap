// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity sql matches snapshot: meal 1`] = `
"CREATE TABLE \\"meal\\" (
  \\"id\\" integer PRIMARY KEY AUTOINCREMENT NOT NULL,
  \\"name\\" varchar NOT NULL,
  \\"date\\" text NOT NULL DEFAULT (CURRENT_TIMESTAMP),
  \\"userId\\" integer,
  \\"macroCarbs\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroProts\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroFats\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroKcal\\" decimal(6, 2) NOT NULL DEFAULT (0),
  CONSTRAINT \\"FK_419ad998c5e3b37a7cce0f872f5\\" FOREIGN KEY (\\"userId\\") REFERENCES \\"user\\" (\\"id\\") ON DELETE NO ACTION ON UPDATE NO ACTION
)"
`;

exports[`entity sql matches snapshot: meal_product 1`] = `
"CREATE TABLE \\"meal_product\\" (
  \\"mealId\\" integer NOT NULL,
  \\"productId\\" integer NOT NULL,
  \\"quantity\\" decimal(4, 1) NOT NULL,
  CONSTRAINT \\"FK_ad97c8c16a4ff39daaf8f464279\\" FOREIGN KEY (\\"mealId\\") REFERENCES \\"meal\\" (\\"id\\") ON DELETE CASCADE ON UPDATE NO ACTION,
  CONSTRAINT \\"FK_769d3e5caceede515715277c784\\" FOREIGN KEY (\\"productId\\") REFERENCES \\"product\\" (\\"id\\") ON DELETE NO ACTION ON UPDATE NO ACTION,
  PRIMARY KEY (\\"mealId\\", \\"productId\\")
)"
`;

exports[`entity sql matches snapshot: product 1`] = `
"CREATE TABLE \\"product\\" (
  \\"id\\" integer PRIMARY KEY AUTOINCREMENT NOT NULL,
  \\"name\\" varchar NOT NULL,
  \\"barcode\\" text,
  \\"unit\\" text NOT NULL DEFAULT ('g'),
  \\"createdAt\\" text NOT NULL DEFAULT (CURRENT_TIMESTAMP),
  \\"updatedAt\\" text NOT NULL DEFAULT (CURRENT_TIMESTAMP),
  \\"isVerified\\" boolean,
  \\"userId\\" integer,
  \\"macroCarbs\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroProts\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroFats\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroKcal\\" decimal(6, 2) NOT NULL DEFAULT (0),
  CONSTRAINT \\"UQ_3e46cda47a776b779fa6faaed10\\" UNIQUE (\\"barcode\\", \\"isVerified\\"),
  CONSTRAINT \\"UQ_c1dd1975fe160d2190161668c52\\" UNIQUE (\\"barcode\\", \\"userId\\"),
  CONSTRAINT \\"UQ_905c0fed2f9ceb88491e43331ec\\" UNIQUE (\\"name\\", \\"isVerified\\"),
  CONSTRAINT \\"CHK_fc8bb6dce3ee11546ec5f9d2bb\\" CHECK (\\"unit\\" IN ('g', 'ml')),
  CONSTRAINT \\"FK_329b8ae12068b23da547d3b4798\\" FOREIGN KEY (\\"userId\\") REFERENCES \\"user\\" (\\"id\\") ON DELETE NO ACTION ON UPDATE NO ACTION
)"
`;

exports[`entity sql matches snapshot: product_images 1`] = `
"CREATE TABLE \\"product_images\\" (
  \\"id\\" integer PRIMARY KEY AUTOINCREMENT NOT NULL,
  \\"url\\" varchar NOT NULL,
  \\"productId\\" integer,
  CONSTRAINT \\"FK_b367708bf720c8dd62fc6833161\\" FOREIGN KEY (\\"productId\\") REFERENCES \\"product\\" (\\"id\\") ON DELETE CASCADE ON UPDATE NO ACTION
)"
`;

exports[`entity sql matches snapshot: product_portions 1`] = `
"CREATE TABLE \\"product_portions\\" (
  \\"productId\\" integer NOT NULL,
  \\"type\\" varchar NOT NULL,
  \\"value\\" integer NOT NULL,
  CONSTRAINT \\"UQ_422bab5f9d32229ab5eec6f1631\\" UNIQUE (\\"productId\\", \\"type\\"),
  CONSTRAINT \\"FK_03c8a652a50d45ef2abcda1db18\\" FOREIGN KEY (\\"productId\\") REFERENCES \\"product\\" (\\"id\\") ON DELETE CASCADE ON UPDATE NO ACTION,
  PRIMARY KEY (\\"productId\\", \\"type\\")
)"
`;

exports[`entity sql matches snapshot: profile 1`] = `
"CREATE TABLE \\"profile\\" (
  \\"id\\" integer PRIMARY KEY AUTOINCREMENT NOT NULL,
  \\"weight\\" tinyint NOT NULL,
  \\"height\\" tinyint NOT NULL,
  \\"age\\" tinyint NOT NULL,
  \\"male\\" boolean NOT NULL,
  \\"weightGoal\\" text NOT NULL,
  \\"userId\\" integer NOT NULL,
  \\"macroNeedsCarbs\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroNeedsProts\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroNeedsFats\\" decimal(6, 2) NOT NULL DEFAULT (0),
  \\"macroNeedsKcal\\" decimal(6, 2) NOT NULL DEFAULT (0),
  CONSTRAINT \\"REL_a24972ebd73b106250713dcddd\\" UNIQUE (\\"userId\\"),
  CONSTRAINT \\"CHK_e13c7952dc8655e48c8f83608a\\" CHECK (
    \\"weightGoal\\" IN ('decrease', 'maintain', 'increase')
  ),
  CONSTRAINT \\"FK_a24972ebd73b106250713dcddd9\\" FOREIGN KEY (\\"userId\\") REFERENCES \\"user\\" (\\"id\\") ON DELETE NO ACTION ON UPDATE NO ACTION
)"
`;

exports[`entity sql matches snapshot: user 1`] = `
"CREATE TABLE \\"user\\" (
  \\"id\\" integer PRIMARY KEY AUTOINCREMENT NOT NULL,
  \\"email\\" text,
  \\"login\\" varchar NOT NULL,
  \\"password\\" varchar NOT NULL,
  \\"createdAt\\" text NOT NULL DEFAULT (CURRENT_TIMESTAMP),
  \\"updatedAt\\" text NOT NULL DEFAULT (CURRENT_TIMESTAMP)
)"
`;
